### PR TITLE
parse_filter parses v0.2 GET filters

### PR DESCRIPTION
**Pull request by csweaver**
This is basically a rewrite of the parse_filter function to parse the GET version of complex filters.

GET /data?obs:tissue=lung&obs:tissue=heart&obs:num_reads=1000,*

### CODE REVIEW

File: server/app/util/filter.py [hunk 1]
```diff
@@ -9,18 +14,27 @@ def _convert_variable(datatype, variable):
...
+    assert datatype in ["boolean", "categorical", "float32", "int32", "string"]
+    if variable is None:
         return variable
+    try:
+        if datatype == "int32":
+            variable = int32(variable)
+        elif datatype == "float32":
+            variable = float32(variable)
+        elif datatype == "boolean":
+            if variable == "false":
```
Comment (ttung) [inline server/app/util/filter.py]: `json.loads(variable)` ?

File: server/app/util/filter.py [hunk 2]
```diff
@@ -9,18 +14,27 @@ def _convert_variable(datatype, variable):
...
+                variable = False
+            elif variable == "true":
+                variable = True
+            else:
+                raise ValueError
```
Comment (ttung) [inline server/app/util/filter.py]: When you raise an exception, you should raise an instance, i.e., `raise ValueError()`.

bonus points for adding an exception message.

File: server/app/util/filter.py [hunk 3]
```diff
@@ -9,18 +14,27 @@ def _convert_variable(datatype, variable):
...
     except ValueError:
```
Comment (ttung) [inline server/app/util/filter.py]: I think you can remove the entire try: except: construct, and just return variable.

File: server/app/util/filter.py [hunk 4]
```diff
@@ -35,36 +49,47 @@ def parse_filter(filter, schema):
...
-        query[key] = {
-            "variable_type": schema[key]["variabletype"],
-            "value_type": schema[key]["type"]
-        }
-        if query[key]["variable_type"] == "categorical":
-            query[key]["query"] = [_convert_variable(query[key]["value_type"], v) for v in value]
-        elif query[key]["variable_type"] == "continuous":
-            value = value[0]
+        axis, annotation = key.split(":", 1)
+        if axis not in [ax for ax in Axis]:
```
Comment (ttung) [inline server/app/util/filter.py]: recommend:

```
try:
    Axis(axis)
except ValueError:
    raise QueryStringError(f"Error: bad axis in query string {axis}")
```

enumtype(value) will look up an enum by value.

File: server/app/util/filter.py [hunk 5]
```diff
@@ -35,36 +49,47 @@ def parse_filter(filter, schema):
...
+            raise QueryStringError(f"Error: bad axis in query string {axis}")
+        if axis not in query:
```
Comment (ttung) [inline server/app/util/filter.py]: consider [collections.defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict) (i would not necessarily adopt it, but you can look at it and see if you think it's right for you)

File: server/app/util/filter.py [hunk 6]
```diff
@@ -35,36 +49,47 @@ def parse_filter(filter, schema):
...
+            query[axis] = {}
+
+        if "annotation_value" not in query[axis]:
```
Comment (ttung) [inline server/app/util/filter.py]: could also use collections.defaultdict here if you want.

File: server/app/util/filter.py [hunk 7]
```diff
@@ -35,36 +49,47 @@ def parse_filter(filter, schema):
     The results is a matrix with the cells the pass the filter and at this point all the genes
     :param filter: flask's request.args
     :param schema: dictionary schema
+    :raises QueryStringError
     :return:
     """
     query = {}
+
     for key in filter:
```
Comment (ttung) [inline server/app/util/filter.py]: fyi, filter is a python keyword.  may be confusing to use it as a variable name.

Comment (ttung) [reply server/app/util/filter.py]: also, what's the type of filter?  can you extract the key + value at the same time like a dict?

File: server/app/util/filter.py (hunk 1 above)
Comment (csweaver) [reply server/app/util/filter.py]: Cool! Didn't realize loads didn't have to take an object/array

File: server/app/util/filter.py (hunk 7 above)
Comment (csweaver) [reply server/app/util/filter.py]: It's ImmutableMultiDict type. The reason I'm not getting the values is that I need to use `getlist` method if it is a categorical type and `get` if it is a numerical type.

File: server/app/util/filter.py (hunk 5 above)
Comment (csweaver) [reply server/app/util/filter.py]: Added that. Good catch, I had meant to use default dict and forgot.

File: server/app/util/filter.py (hunk 4 above)
Comment (csweaver) [reply server/app/util/filter.py]: Done. I had some trouble figuring out the best way to check that, thanks

File: server/app/util/filter.py (hunk 3 above)
Comment (csweaver) [reply server/app/util/filter.py]: Fixed.

File: server/app/util/filter.py (hunk 2 above)
Comment (csweaver) [reply server/app/util/filter.py]: good point. moot now, though

**Pull request merged**